node/impl/full/actor_events.go

```diff
@@ -0,0 +1,226 @@
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/events/filter"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type ActorEventAPI interface {
+	GetActorEvents(ctx context.Context, filter *types.ActorEventFilter) ([]*types.ActorEvent, error)
+	SubscribeActorEvents(ctx context.Context, filter *types.SubActorEventFilter) (<-chan *types.ActorEvent, error)
+}
+
+var (
+	_ ActorEventAPI = *new(api.FullNode)
+	_ ActorEventAPI = *new(api.Gateway)
+)
+
+type ActorEventHandler struct {
+	EventFilterManager   *filter.EventFilterManager
+	MaxFilterHeightRange abi.ChainEpoch
+	Chain                *store.ChainStore
+}
+
+var _ ActorEventAPI = (*ActorEventHandler)(nil)
+
+type ActorEventsAPI struct {
+	fx.In
+	ActorEventAPI
+}
+
+func (a *ActorEventHandler) GetActorEvents(ctx context.Context, filter *types.ActorEventFilter) ([]*types.ActorEvent, error) {
+	if a.EventFilterManager == nil {
+		return nil, api.ErrNotSupported
+	}
+
+	params, err := a.parseFilter(filter)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a temporary filter
+	f, err := a.EventFilterManager.Install(ctx, params.MinHeight, params.MaxHeight, params.TipSetCid, filter.Addresses, filter.Fields, false)
+	if err != nil {
+		return nil, err
+	}
+
+	evs, err := getCollected(ctx, f)
+	_ = a.EventFilterManager.Remove(ctx, f.ID())
+	return evs, err
+}
+
+type filterParams struct {
+	MinHeight abi.ChainEpoch
+	MaxHeight abi.ChainEpoch
+	TipSetCid cid.Cid
+}
+
+func (a *ActorEventHandler) parseFilter(f *types.ActorEventFilter) (*filterParams, error) {
+	if f.TipSetCid != nil {
+		if len(f.FromEpoch) != 0 || len(f.ToEpoch) != 0 {
+			return nil, fmt.Errorf("cannot specify both TipSetCid and FromEpoch/ToEpoch")
+		}
+
+		return &filterParams{
+			MinHeight: 0,
+			MaxHeight: 0,
+			TipSetCid: *f.TipSetCid,
+		}, nil
+	}
+
+	from := f.FromEpoch
+	if len(from) != 0 && from != "latest" && from != "earliest" && !strings.HasPrefix(from, "0x") {
+		from = "0x" + from
+	}
+
+	to := f.ToEpoch
+	if len(to) != 0 && to != "latest" && to != "earliest" && !strings.HasPrefix(to, "0x") {
+		to = "0x" + to
+	}
+
+	min, max, err := parseBlockRange(a.Chain.GetHeaviestTipSet().Height(), &from, &to, a.MaxFilterHeightRange)
+	if err != nil {
+		return nil, err
+	}
+
+	return &filterParams{
+		MinHeight: min,
+		MaxHeight: max,
+		TipSetCid: cid.Undef,
+	}, nil
+}
+
+func (a *ActorEventHandler) SubscribeActorEvents(ctx context.Context, f *types.SubActorEventFilter) (<-chan *types.ActorEvent, error) {
+	if a.EventFilterManager == nil {
+		return nil, api.ErrNotSupported
+	}
+
+	params, err := a.parseFilter(&f.Filter)
+	if err != nil {
+		return nil, err
+	}
+
+	fm, err := a.EventFilterManager.Install(ctx, params.MinHeight, params.MaxHeight, params.TipSetCid, f.Filter.Addresses, f.Filter.Fields, false)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(chan *types.ActorEvent, 25)
```

Comment:
Hm. Given that we're not batching by block, this buffer is _tiny_ (likely way too tiny) to simply disconnect the reader if we get stuck.

We likely need some leniency. E.g., the user gets to block for a few milliseconds at a time before we kill the reader. Or, even, the user gets some form of "budget" where they get to block for, say, 10% of the time? I.e., if they get to block for 3 seconds every 30s, they'll never _really_ fall behind.

Or maybe just a larger buffer? I'm just worried about this API may be completely unusable in practice (given any amount of RPC latency).

Reply:
A somewhat complex algorithm that should "work" (after prefill) is:

1. Read from the event channel into a slice.
2. Record the height of the event from the _newest_ tipset observed. Don't make reduce this height if we hit a revert.
3. If the height increases and the slice isn't empty, disconnect the reader.

Except when the node is "catching up" (probably not something we care about anyways), this will always give the reader a full block time to receive events but won't let them fall behind.

However, the downside is memory. _Ideally_ the event filter manager would return _batches_ of events (the same slice for every subscriber).

Reply:
In terms of implementation, one usually does this as follows:

```go
var buffer []*types.ActorEvent
for ctx.Err() == nil {
    var ev *types.ActorEvent
    var ok bool
    if len(buffer) > 0 {
        select {
        case ev, ok = <-in:
        case out<-buffer[0]:
            buffer[0] = nil
            buffer = buffer[1:]
            continue
        case <-ctx.Done():
            return
        }
    } else {
        select {
        case ev, ok = <-in:
        case <-ctx.Done():
            return
        }
    }
    if !ok {
        return
    }
    // check if we need to disconnect the client because they've fallen behind.
    // ...
    
    // Otherwise, add the event to the buffer
    buffer = append(buffer, ev)
}
```

Reply:
But... this is an implementation thing. We _can_ fix this in a followup patch (no API changes).

Reply:
For pre-filling, we'll likely just need some kind of minimum rate (e.g., one event per millisecond, maybe averaged out over some period of time).

Reply:
@Stebalien a case I've hit that I'm unsure about in terms of what patterns to follow with existing APIs - what about the case where we can clearly identify a halting condition, do we close the channel? If I specify a `ToHeight` that's 100 epochs in the future (or even do a subscribe that has the range entirely historical), does the API terminate itself? It seems like _yes_ is the right answer to this, at least that's what I'd expect.

Reply:
> If I specify a ToHeight that's 100 epochs in the future

You mean on subscribe? Ideally you'd terminate when hitting the target height. Although it would nice to be able to signal success/error... i think that's a limitation of our JSON-RPC API more than anything.